src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -284,6 +298,82 @@ private class ScopeTreeBuilder : BoundTreeWalkerWithStackGuardWithoutRecursionOn
                 private readonly HashSet<MethodSymbol> _methodsConvertedToDelegates;
                 private readonly DiagnosticBag _diagnostics;
 
+                private static readonly ObjectPool<List<Scope>> s_scopeListObjectPool = new ObjectPool<List<Scope>>(() => new List<Scope>());
+
+                /// <summary>
+                /// For every label visited so far, this dictionary maps to a list of all scopes either visited so far, or currently being visited,
+                /// that are both after the label, and are on the same level of the scope tree as the label.
+                /// </summary>
+                private readonly PooledDictionary<LabelSymbol, List<Scope>> _scopesAfterLabel = PooledDictionary<LabelSymbol, List<Scope>>.GetInstance();
+
+                private static readonly ObjectPool<List<LabelSymbol>> s_labelListObjectPool = new ObjectPool<List<LabelSymbol>>(() => new List<LabelSymbol>());
+
+                private static readonly ObjectPool<Stack<List<LabelSymbol>>> s_stackObjectPool = new ObjectPool<Stack<List<LabelSymbol>>>(() => new Stack<List<LabelSymbol>>());
+
+                /// <summary>
+                /// Contains a list of the labels visited so far for each scope. 
+                /// The stack represents the chain of scopes from the root scope to the current scope.
+                /// </summary>
+                private readonly Stack<List<LabelSymbol>> _labelsInScope = s_stackObjectPool.Allocate();
```

Comment:
This should be the same thing as walking from the current scope to the root through the parent nodes

Reply:
Never mind, the comment confused me. Why do we need this?

Reply:
It's used when moving to a child scope, to add the child scope to the _scopesAfterLabelDictionary for all those labels. I'll add a cref in the comments.